### ISSUE TITLE
Datatype syntax for contracts

### ISSUE THREAD

**Issue by yannham** [tweag/nickel#445]
**Is your feature request related to a problem? Please describe.**
Parameterizing a contract by one or several other contracts is something natural and useful. This is the counterpart of parameterizing an algebraic datatype by other types in static type systems. The simplest example is probably the nullable contract:

```
let Nullable = fun contract label value =>
  if value == null then null
  else contract label value
```

However, using it is a bit clumsy. One would like to use it as a datatype, like `let foo | #Nullable Str = "foo"`, but it accepts a value (representing a contract), not a type. Thus we need to pass it a contract, like `let foo | #(Nullable str) = ...` where `str` is the builtin contract for strings. This is a bit confusing, and makes it hard or even impossible to use it with more advanced types (the equivalent of `#Nullable (forall a b. a -> Num -> {foo: b})` is highly non trivial to write by hand, if possible).

**Describe the solution you'd like**
I would like the syntax to allow applying a custom contracts to a type: `#Foo Bar`. The semantics is that it would be equivalent to `#(Foo bar)` where `bar` is the contract derived by the interpreter from the type `Bar`. Typically, for custom contracts that would be `#Foo #Bar ~ #(Foo Bar)`. But it allows to use any type whatsoever, including record types, list, arrow types, etc: `#Foo (List #Bar)`. This would give an easy way to have algebraic-datatype-like contracts, parametrized by other contracts.

Note that contracts are sometimes parametrized by values. This is the contract counterpart of dependent types:

```
let GreaterThan = fun base label value => value >= base || contracts.blame label
```

This will still require parentheses, and won't clash with the proposed syntax: `let foo | #(GreaterThan 2) = 5`. The proposed syntax is just a convenient special case when the value the contract depends on is another contract.

**Comment by Radvendii** [comment]
Two questions:
1) What happens if the contract takes both another contract and a value as parameters?
2) Should types just always be useable in this way in contexts that expect contracts?

**Comment by yannham** [comment]
> What happens if the contract takes both another contract and a value as parameters?

Then you have to order your arguments to take advantage of this datatype syntax. Contracts first and then more general values. Note that you can always fallback to the current situation and pass everything as contracts if you really want to have shuffled arguments, but you probably shouldn't.

> Should types just always be useable in this way in contexts that expect contracts?

If I understand your question correctly, yes. Actually, things like `Num` and `#Foo` are technically types. The thing is there a correspondence in the two directions: to any contract represented as a value (function) like `foo` corresponds an opaque type `#foo`, and to any type like `Num` corresponds a contract that can be generated by the interpreter's `open_contract()`. The thing is while the user can easily go in one direction (put a `#` as a prefix to a contract), the other is implicit. `open_contract()` is called by the interpreter e.g. to generate the check corresponding to writing `someVal | Num`, but you can't access it as a language construct in general.

One of the point of this syntax is to make it easy to pass both custom contracts and builtin contracts as arguments to other contracts uniformly, by working on the side of types instead of the side of terms.

**Comment by yannham** [comment]
> Why did we need the # syntax to begin with? I remember that #… was chosen arbitrarily as an easy-to-parse placeholder. I remember that there was an actual parsing issue, but I don't remember what it was.

I don't know the original motivation to be honest, it has been there before me. It's probably easier to parse and for differentiating built-in types from custom contracts. But currently I find it useful to have non-trivial inline contracts, separating term syntax and type syntax, e.g. typically for "dependent contracts":

```
let GreaterThan = fun base label value => if value > base then value else contracts.blame label in
let Nullable = ... in
let OneOf = fun list label value => if lists.elem value list then value else contracts.blame label in

let foo | #(GreaterThan 2) = 5
let bar | #Nullable #{bar | Str | default = "", ..} = {}
let baz | #(OneOf ["foo", "bar", 2, null]) = 2
```

That is, to allow something else than just an identifier as a contract. In the codebase I'm converting, I use this quite a lot, as it would be very heavy to introduce an identifier for each and every small field contracts (especially bound constraints) I want to use.

Changing the syntax has been touched on before, though (e.g. in #318). This is something I want to get into soon: gather several of those pending technically not hard but practically important choices that have been spread out and often already discussed in different issues (but not fixed either), and have a standardization meeting where we decide all of those, including the syntax of custom contracts.

However, this is slightly orthogonal the proposal, as whatever syntax we choose in the end, this issue is about giving a semantics to type application when the LHS is a custom contract.

**Comment by aspiwack** [comment]
> However, this is slightly orthogonal the proposal, as whatever syntax we choose in the end, this issue is about giving a semantics to type application when the LHS is a custom contract.

I'm not so sure that they are. There seem to be a pretty large design issue around types and contracts. And these may need to be considered as a whole. At least to some degree.

**Comment by yannham** [comment]
Solved/superseded by #573.
